### ISSUE TITLE
Ja4 Supporting

### ISSUE THREAD

**Issue by yarosman** [phuslu/nginx-ssl-fingerprint#42]
There exist updates in TLS client fingerprinting technology 
https://github.com/FoxIO-LLC/ja4/tree/main

**Comment by phuslu** [comment]
Thanks! I admit that ja4 do some enhancements/improvements than ja3, but I'd like wait it to be mature
I think that bigtech (e.g. cloudflare)  will merge/support ja4 variants or raise another alternative for it.
Let's keep eyes on it.

**Comment by phuslu** [comment]
a bit related/similar with https://github.com/phuslu/nginx-ssl-fingerprint/issues/13

**Comment by yarosman** [comment]
Chrome randomizes extensions, Firefox will randomize in the future (I read somewhere)) ) - therefore without ordering, we have cases when on each page reloading we will get different ja3, and ja4 fixes it

**Comment by ne4u** [comment]
> Currently I dont like ja4 because it sorted extension -- randomize extensions is a feature of new chrome -- we could distinguish/detect bot on top of it, so we shall not sort. Rest parts of ja4 is LGTM.

I agree.  here's why:  if you have a client that identifies as a current version of chrome but does not have a random extension list on subsequent requests, you know it's a bot.  An unsorted list can always be sorted later.

But, I also agree that the hash by itself is less useful.  Just out of curiosity what happens when FIPS mode is enabled on the system or during compile time since md5 function is removed?  It seems to me the hash function can be removed and any hash functionality could be done by whatever process consumes the fingerprint data.

**Comment by phuslu** [comment]
Cloudflare added the support of JA4 in its enterprise plan, it's my turn to follow it now.

https://developers.cloudflare.com/bots/concepts/ja3-ja4-fingerprint/

**Comment by HanadaLee** [comment]
> Cloudflare added the support of JA4 in its enterprise plan, it's my turn to follow it now.
> 
> https://developers.cloudflare.com/bots/concepts/ja3-ja4-fingerprint/

Is there any progress on supporting ja4 in this project?

**Comment by phuslu** [comment]
Oh, I spent too much time in my go projects, let me pick up this in recent days.

**Comment by dedok** [comment]
Btw I need this feature also. We can collaborate here :)

**Comment by dolgovas** [comment]
I'm waiting for this feature too

**Comment by HanadaLee** [comment]
For those who need ja4, this fork may be worth considering: https://github.com/hnakamur/nginx-ssl-fingerprint

I also made further modifications based on this, mainly to avoid using variables with the $http_ prefix, so as to avoid confusion with HTTP request header variables: https://github.com/HanadaLee/ngx_ssl_fingerprint_module

**Comment by dolgovas** [comment]
@HanadaLee may be you'll create PR?

**Comment by HanadaLee** [comment]
> [@HanadaLee](https://github.com/HanadaLee) may be you'll create PR?

My branch has many break changes and is not suitable for merging into this repository.

**Comment by phuslu** [comment]
ja4 is landed on master branch, can give a try now.

Once I completed the whole performance optimization, I will release a new tag.

**Comment by phuslu** [comment]
Released [v1.0.5](https://github.com/phuslu/nginx-ssl-fingerprint/releases/tag/v1.0.5) for ja4 support. Thanks all. Bug reports and Pull requests are welcome.
